src/main/java/org/opentripplanner/routing/trippattern/TripTimes.java

```diff
@@ -360,6 +372,58 @@ public int getDropoffType(int stop) {
         return dropoffs[stop];
     }
 
+    /**
+     * Calculates the earliest pickup time based on a stop's pickup booking rule (GTFS-Flex v2.1)
+     * @param stop - iterator in the StopTimes collection
+     * @param currTime - current time
+     * @param sd - service day
+     * @return time
+     */
+    public long getEarliestPickupBookingTime (int stop, long currTime, ServiceDay sd) {
+        BookingRule pickupBookingRule = pickupBookingRules[stop];
+        int bookingRuleType = pickupBookingRule.getType();
+        switch (bookingRuleType) {
+            case 0:
+                // Real-time booking only
+                return currTime;
+            case 1:
+                // Up to same-day booking, with advance notice
+                return (long) (currTime + Math.round(pickupBookingRule.getPriorNoticeDurationMin() * 60.0));
+            case 2:
+                // Up to prior day(s) booking
+                int priorNoticeStartDay = pickupBookingRule.getPriorNoticeStartDay();
+                int priorNoticeStartTime = pickupBookingRule.getPriorNoticeStartTime();
+                return sd.time((int) Math.round((priorNoticeStartDay * 24.0 + priorNoticeStartTime) * 60.0));
```

Comment:
This is something I am not sure if it produces a number of seconds from now (`currTime`) till service date in `priorNoticeStartDay` days + `priorNoticeStartTime` hours. 🤔 There are too many factors in play: service calendar, timezone, current time and the time mappers.

Reply:
I guess the only way forward here is to have unit tests that cover different types of booking rules 🤔

Reply:
The `priorNoticeStartDay` should probably operate on days, converting to ServiceDate and back, rather that 24 hour increments, to account for switching to an from DST.